Update CONTRIBUTING.md and pull_request_template.md to provide more information to developers of features

cmcantalupo opened the pull request:
- Fixes #1878 from github issues.
- Feature development was not really covered in CONTRIBUTING.md only a brief note about opening a request for a feature.
- Provides more complete list of requirements for PRs in the pull request template.

- Required details
    + <details> <summary>Documentation fix</summary>

        - The modification was to developer documentation
        - The change was an addition
        - The reason for the change was lack of clarity on PR requirements and process
        - The documentation component that changed were the github defined files CONTIRBUTING.md and pull_request_template.md

      </details>

cmcantalupo committed:
Add issue templates and update CONTRIBUTING.md

- Add github issue templates for bugs feature requests,
  stories, tasks and technichal debt
- Update CONTRIBUTING with more information about how to
  make a change to the repo.
- Update the pull request template
  + Add a section for each type of PR
  + Required steps for each type are described in drop down
  + Expectations are listed in the template

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on pull_request_template.md:
```diff
@@ -1,4 +1,97 @@
+              if used in C++
+        - [ ] All existing feature continue to work when dependency is
+              not available
+      </details>
```
ws

bgeltz commented on pull_request_template.md:
```diff
@@ -1,4 +1,98 @@
+              system in this or a prior PR
+        - [ ] Dependency has one or more last-level-abstraction classes
+              if used in C++
+        - [ ] All existing feature continue to work when dependency is
```
```suggestion
        - [ ] All existing features continue to work when dependency is
```

bgeltz commented on pull_request_template.md:
```diff
@@ -1,4 +1,98 @@
+
+    + <details> <summary>Bug fix</summary>
+
+        - [ ] Details about the bug are located the above bug report
```
```suggestion
        - [ ] Details about the bug are located in the above bug report
```

bgeltz commented on pull_request_template.md:
```diff
@@ -1,4 +1,98 @@
+
+    + <details> <summary>New methods added to a C++ class</summary>
+
+        - [ ] Unit tests added for new methods
+        - [ ] Mocks objects derived from class have methods added
```
```suggestion
        - [ ] Unit tests added for the new methods
        - [ ] Mock objects derived from class have methods added
```

cmcantalupo commented on .github/ISSUE_TEMPLATE/story.md:
```diff
@@ -0,0 +1,19 @@
+A clear and concise description of any alternative solutions or features you've considered.
+
+**Additional context**
+Add any other context about the feature request here.
```
```suggestion
Add any other context about the story here.
```

cmcantalupo committed:
Convert CONTRIBUTING from md to ReST format and split it up

- Split CONTRINBUTING into two parts, developer doc moved to sphinx page.
  - Leave all instructions that are not specifically about changes to
    software.
  - Move developer specific instructions and create a link to them.
- Get rid of all caps headers.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove empty guides from the index

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented:
Renderings of the contributing and development documentation from this PR can be found here:

https://github.com/cmcantalupo/geopm/blob/issue-1878/CONTRIBUTING.rst
https://cmcantalupo.github.io/geopm.github.io/geopmdpy/contrib.html
https://cmcantalupo.github.io/geopm.github.io/geopmdpy/devel.html

The devel.html is drawn directly from the CONTRIBUTING.md and should be updated to include information about the service build.  The pull request template or issue templates can be demoed by starting to make issues or pull requests to my fork of GEOPM.

cmcantalupo commented:
The rendered pull request template can be seen here:
https://github.com/cmcantalupo/geopm/blob/issue-1878/pull_request_template.md

And an example of the issue creation interface can be seen here (click through to see template rendered for each issue type):
https://github.com/cmcantalupo/geopm/issues/new/choose

cmcantalupo committed:
Move PR template to task templates plus many improvements based on feedback

- Add license check to list for adding external dependency
- Add a section to the pull request template for test improvements
- Move PR template data into issue templates
- Add paragraph to CONTRIBUTING about issue requirement for PR
- Update CONTRIBUTING to reflect the enhanced issue template
- Address some review comments and add more process documentation
- "Task" templates are now called "Change" templates
- CONTRIBUTING.rst now is explicit about the 1:1 mapping from
  PRs to Change template tasks.
- Clarify some of the text in issues based on Brad's feedback.
- Reference contributing in pull request template.
- Add feature/story selection to changes issue templates
- Add more links to contributing documentation in templates
- Add suggestions from Brad about python deps

Co-authored-by: Brad Geltz <brad.geltz@intel.com>
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd commented on pull_request_template.md:
```diff
@@ -1,4 +1,124 @@
+        - [ ] All dependency classes have already been merged
+        - [ ] All public methods have doxygen comments describing
+              behavior, inputs, outputs and exceptions raised
+        - [ ] Implementation file has `#include "config.h"`
```
This might be something we could automate, either by guaranteeing at compile time or by verifying at test time. Opened #1890 as a tech-debt issue to automate this.

cmcantalupo commented:
Working on transferring pull request template data into issue templates.  I will reopen for review when completed

bgeltz commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,25 @@
+I tried to [...] and I expected [...] instead [...]
+
+**GEOPM version**
+E.g. output from `geopmread --version`
```
We should be explicit about what we want here.  It shouldn't be left up to interpretation.
```suggestion
Output from `geopmread --version`:
<OUTPUT>

Output from `geopmagent --version`:
<OUTPUT>
```

cmcantalupo replied:
The issue is that if someone is only using the service, they may not have even built or installed geopmagent.  They should always have geopmread in their path, maybe we should just use that throughout the documentation?

bgeltz commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,25 @@
+A clear and concise description of what actually happened
+
+**Error messages produced**
+If possible use `configure --enable-debug ...` for more verbose error messaging.
```
```suggestion
If possible use `configure --enable-debug ...` (or `GEOPM_GLOBAL_CONFIG_OPTIONS="--enable-debug" ./build.sh` if you're using the build script) for more verbose error messaging.
```

bgeltz commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,25 @@
+If possible use `configure --enable-debug ...` for more verbose error messaging.
+
+**Additional context**
+Add any other context about the problem here.
```
```suggestion
Add any other context about the problem here (e.g. the architecture of the system under test, the number of nodes, MPI ranks, and OMP threads used, etc.).
```

bgeltz commented on .github/ISSUE_TEMPLATE/feature_request.md:
```diff
@@ -0,0 +1,19 @@
+
+---
+**Is your feature request related to a problem? Please describe.**
+A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]
```
```suggestion
A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]
If applicable, a link to any existing bug reports that are related.
```

cmcantalupo replied:
If there are related bugs, should the issue just be a task_bug_fix.md template?

bgeltz commented on .github/ISSUE_TEMPLATE/feature_request.md:
```diff
@@ -0,0 +1,19 @@
+---
+name: Feature request
+about: Suggest an idea for this project
+title: As a [...] I would like [...]
```
Does it make sense to include pointers to wherever the descriptions of the roles is nearby this text?
Maybe in the top of the body below?

bgeltz commented on .github/ISSUE_TEMPLATE/story.md:
```diff
@@ -0,0 +1,19 @@
+---
+name: Story
+about: Small problem that can be solved within a sprint
+title: As a [...] I would like [...]
```
Does it make sense to include pointers to wherever the descriptions of the roles is nearby this text?
Maybe in the top of the body below?

bgeltz commented on .github/ISSUE_TEMPLATE/story.md:
```diff
@@ -0,0 +1,19 @@
+A clear and concise description of any alternative solutions or features you've considered.
+
+**Additional context**
+Add any other context about the story here.
```
```suggestion
Add any other context about the story here (e.g. links to the feature request, bug report, or related issue).
```

bgeltz commented on .github/ISSUE_TEMPLATE/task_bug_fix.md:
```diff
@@ -0,0 +1,15 @@
+labels: ''
+assignees: ''
+
+---
```
Maybe a reminder here that we do not assign points to bugs.  I wonder if we can do an explicit estimate of 0 or similar.

bgeltz commented on .github/ISSUE_TEMPLATE/task_python_add_class.md:
```diff
@@ -0,0 +1,24 @@
+      behavior, inputs, outputs and exceptions raised
+- [ ] Isolate external dependencies in wrapping classes or modules if
+      doing so enables better testing.
+- [ ] [...]
```
```suggestion
- [ ] The new class is fully compliant with Python 3.  Python 2 is not supported.
- [ ] If an external module dependency was introduced, the new module is added to the requirements.txt and is available via PyPI (pip).
- [ ] [...]
```

bgeltz commented on .github/ISSUE_TEMPLATE/task_python_add_methods.md:
```diff
@@ -0,0 +1,19 @@
+- [ ] Unit tests do not refer to private methods or members
+- [ ] New public methods have __doc__ string comments describing
+      behavior, inputs, outputs and exceptions raised
+- [ ] [...]
```
```suggestion
- [ ] The new class is fully compliant with Python 3.  Python 2 is not supported.
- [ ] If an external module dependency was introduced, the new module is added to the requirements.txt and is available via PyPI (pip).
- [ ] [...]
```

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -3,4 +3,115 @@ Guide for GEOPM Developers
 ==========================
 
 These are instruction for a developer that would like to modify the
```
```suggestion
These are instructions for a developer that would like to modify the
```

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -3,4 +3,115 @@ Guide for GEOPM Developers
 
 These are instruction for a developer that would like to modify the
-geopm repository.
+source code in the geopm repository.
```
```suggestion
source code in the GEOPM repository.
```

Should we scrub all this docu for geopm -> GEOPM changes?

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -3,4 +3,115 @@ Guide for GEOPM Developers
+Test Instructions
+-----------------
+
+To launch the GEOPM unit tests, run the following command in the geopm
+build directory after running configure:
```
Which GEOPM build directory?  The base dir or the service dir?

The build script also supports this for running both directories:
```
GEOPM_RUN_TESTS=yes ./build.sh
```

cmcantalupo replied:
This is just moved text that used to be in README.md

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -3,4 +3,115 @@ Guide for GEOPM Developers
+
+.. code-block::
+
+   ./configure --disable-mpi
```
Only applicable for the base dir.  Also, might want to include the disable fortran flag.

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -3,4 +3,115 @@ Guide for GEOPM Developers
+.. code-block::
+
+   ./configure --disable-mpi
+   make
+   make check
```
```suggestion
   make
   make checkprogs  # OPTIONAL
   make check
```

Also what about `-j`?

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -3,4 +3,115 @@ Guide for GEOPM Developers
+
+http://ltp.sourceforge.net/coverage/lcov.php
+
+The GEOPM build must be configured with the "--enable-coverage" option.  Then
+simply run
```
There's more to it than this.

1. You have to build with GNU: 
```
$ source ${HOME}/geopm/integration/config/gnu_env.sh
$ GEOPM_SKIP_COMPILER_CHECK=yes source ${HOME}/geopm/integration/config/build_env.sh
$ GEOPM_SKIP_COMPILER_CHECK=yes GEOPM_SKIP_INSTALL=yes GEOPM_GLOBAL_CONFIG_OPTIONS="--enable-coverage" ./build.sh
```
2. You have to be using a very recent version of lcov if you're using GCC 9 or later: https://github.com/linux-test-project/lcov/issues/58.  For me, that amounted to a source build of lcov because the version available in the zypper repos is too old.
3. If you don't use the build script, you'll have to configure/build the dir you want (either base or service).
4. The coverage make target still has shortcomings that need to be addressed, namely that there's no baseline capture before `make check` is run that would capture files that have 0 coverage.  Also, the `coverage` target is setup to do separate webpages now: one if you configure and run out of the service dir, and one if you do the same out of the base dir.  We should have a separate sprint issue about resolving some of this stuff maybe?

cmcantalupo commented on .github/ISSUE_TEMPLATE/task.md:
```diff
@@ -0,0 +1,14 @@
+---
+name: Task - [Empty template]
```
PR - [Empty template]

bgeltz commented on .github/ISSUE_TEMPLATE/bug_report.md:
````diff
@@ -10,16 +10,36 @@ assignees: ''
+run the following command:
+
+```bash
+GEOPM_GLOBAL_CONFIG_OPTIONS="$GEOPM_GLOBAL_CONFIG_OPTIONS --enable-debug" \
````
```suggestion
GEOPM_GLOBAL_CONFIG_OPTIONS="--enable-debug" \
```

Nothing else is setting this env option presently.  Not sure what else would be.  This is the syntax that the help text of the build script recommends.

bgeltz commented on CONTRIBUTING.rst:
```diff
@@ -67,10 +60,23 @@ message is appreciated, but not required.  All pull requests to the
+
+Each change to the repository is identified with an issue created
+using one of the issues templates that begins "Change - ".  The
+"Change" issue should reference the "Bug report," "Reature request,"
```
```suggestion
"Change" issue should reference the "Bug report," "Feature request,"
```

Zoinks.  Like hey Scoob...

cmcantalupo replied:
hehe
